### PR TITLE
HRJS-80 - Connection to not-configured cache fails without closing it

### PR DESCRIPTION
**Pull request by andyuk1986**

### CODE REVIEW

File: lib/infinispan.js [hunk 1]
```diff
@@ -228,6 +228,14 @@
             .then(function() {
               logger.debugf('Started Infinispan %s', client);
               return client; // return client
+            }, function(reason){
```
Comment (galderz) [inline lib/infinispan.js]: `catch()` is preferred and consistent with the rest of the codebase.

```
.then()
.catch(function(error) { ... })
```

File: lib/infinispan.js [hunk 2]
```diff
@@ -228,6 +228,14 @@
...
+              transport.disconnect();
+
+              var error = reason;
+              if (typeof reason == "string") {
```
Comment (galderz) [inline lib/infinispan.js]: Why is this needed exactly? Can't you just throw the error you got?

File: lib/io.js [hunk 3]
```diff
@@ -654,7 +654,7 @@
       disconnect: function() {
         logger.debugf('Disconnect transport');
         emitter.removeListener('router', onRouter);
-        return router.disconnect();
+        return (router ? router.disconnect() : null);
```
Comment (galderz) [inline lib/io.js]: Not sure router could be null in this case. Even if ping returns an error, if the topology has been updated, it's updated. Even on first connection, transport.connect() method should set a fixed router as soon as one of the connections succeds.

**Comment by galderz** [comment]
@andyuk1986 Thanks for the fix, I have some doubts about some of the parts. If you can provide more details that'd be great. Ping me if in doubt.

File: lib/infinispan.js (hunk 1 above)
Comment (andyuk1986) [reply lib/infinispan.js]: Changed according to comment.

File: lib/infinispan.js (hunk 2 above)
Comment (andyuk1986) [reply lib/infinispan.js]: Actually when I was rethrowing the error, I had some issues with ssl tests, but now it is fixed. Anyway, removed this part and just rethrowing the error.

File: lib/io.js (hunk 3 above)
Comment (andyuk1986) [reply lib/io.js]: This line change is a side-effect of the previous change (transport.disconnect() in infinispan.js). 
1. the transport.disconnect() in catch clause in infinispan.js is added as a solution for this bug, because when the client tries to connect to not-configured cache, it gives an error, the transport is not closed and the connection hangs.
2. on the other hand, with this solution some of the ssl  tests start to fail with message: 
Error: [Cannot read property 'disconnect' of undefined] is not any of: [CERT_SIGNATURE_FAILURE,self signed certificate]
and this error is thrown on router.disconnect() in io.js file. That's why I have added the null check.

**Comment by andyuk1986** [comment]
@galderz Thanks a lot for looking into this Galder. I have fixed your suggestions, also have replied to your comments.

File: lib/io.js (hunk 3 above)
Comment (galderz) [reply lib/io.js]: Oh I see, the SSL errors are client side errors that can happen on startup. In this case, I'd make the `disconnect()` function clearer of what it does. So, I'd check with `f.existy(router)` when `disconnect()` is called and the log accordingly. It can be confusing for the log message to say `Disconnect transport` when you actually don't do any disconnection.

Comment (galderz) [reply lib/io.js]: Also, always use `f.existy()` to check if a value is undefined or not. Also, better than returning `null`, return a Promise completed with empty, e.g. `Promise.resolve()`

Comment (andyuk1986) [reply lib/io.js]: Thanks a lot Galder for your comments. I have fixed this part of code as you have suggested.

File: spec/infinispan_local_spec.js [hunk 4]
```diff
@@ -81,6 +81,15 @@ describe('Infinispan local client', function() {
     .catch(t.failed(done))
     .finally(done);
   });
+  it('fails when non-configured cache is accessed', function(done) {
+      t.client(t.local, {cacheName: 'unknownCache'})
+          .then(function() {
+              done(new Error('Exception should be thrown while accessing not-configured cache.'));
+         }).catch(function(error) {
+              expect(error).toBe("org.infinispan.server.hotrod.CacheNotFoundException: Cache with name 'unknownCache' not found amongst the configured caches");
+              t.failed(done);
```
Comment (galderz) [inline spec/infinispan_local_spec.js]: Should be just `done()`. This is the expected path. As a result, `.finally(done)` is not needed.

**Comment by galderz** [comment]
@andyuk1986 Still one comment to deal with.

File: spec/infinispan_local_spec.js (hunk 4 above)
Comment (andyuk1986) [reply spec/infinispan_local_spec.js]: @galderz Fixed.

**Comment by andyuk1986** [comment]
Hi @galderz Sorry for my late response. I have fixed the last comment.

**Comment by galderz** [comment]
Thx @andyuk1986
